src/callbacks_step/amr.jl

```diff
@@ -124,6 +128,18 @@ function initialize!(cb::DiscreteCallback{Condition,Affect!}, u, t, integrator)
   amr_callback = cb.affect!
   semi = integrator.p
 
+  mesh, _, _, _ = mesh_equations_solver_cache(semi)
+
+  if mesh isa ParallelP4estMesh
+    if cb.affect!.p4est_partition_allow_for_coarsening != mesh.p4est_partition_allow_for_coarsening
+      mesh.p4est_partition_allow_for_coarsening = cb.affect!.p4est_partition_allow_for_coarsening
+      if mpi_isroot()
+        @info "The mesh attribute `p4est_partition_allow_for_coarsening` was changed to `$(cb.affect!.p4est_partition_allow_for_coarsening)` by the `AMRCallback`."
+      end
+      partition!(mesh)
+    end
+  end
+
```

Comment:
```suggestion
  mesh, equations, solver, cache = mesh_equations_solver_cache(semi)

  if mesh isa ParallelP4estMesh
    if cb.affect!.p4est_partition_allow_for_coarsening != mesh.p4est_partition_allow_for_coarsening
      mesh.p4est_partition_allow_for_coarsening = cb.affect!.p4est_partition_allow_for_coarsening
      if mpi_isroot()
        @info "The mesh attribute `p4est_partition_allow_for_coarsening` was changed to `$(cb.affect!.p4est_partition_allow_for_coarsening)` by the `AMRCallback`."
      end
      global_first_quadrant = unsafe_wrap(Array, unsafe_load(mesh.p4est).global_first_quadrant, mpi_nranks() + 1)
      old_global_first_quadrant = copy(global_first_quadrant)
      partition!(mesh)
      rebalance_solver!(integrator.u, mesh, equations, solver, cache, old_global_first_quadrant)
      reinitialize_boundaries!(semi.boundary_conditions, cache)
    end
  end

```

Unfortunately, I missed something really important here. When the AMR callback is initialized, the cache and the initial solution already exist and thus it is not sufficient to merely repartition the mesh, which I initially suggested. The other data structures must be adjusted as well.

The suggested change above is based on the load balancing code [here](https://github.com/trixi-framework/Trixi.jl/blob/main/src/callbacks_step/amr.jl#L431-L440), I'm just not sure if the use of `integrator.u` is correct here since there is also some other `u` passed as an argument to the `initialize!` function. However, `integrator.u` is used below too. :shrug:

Reply:
Good point! That might need an additional `u_modified!(integrator, true)` after modifying `integrator.u`.

Reply:
@sloede I agree, the implementation ended up being more complex than I anticipated. Setting the default value to `true` and not overriding anything when AMR is used would indeed be a lot simpler to implement and people could still get the performance benefits by adjusting it manually when constructing the mesh.

Reply:
That sounds indeed reasonable to me given the complexity this PR gets.

Reply:
I set the default value to `true` and removed overriding the value in the `AMRCallback` in 517114d.

Reply:
Do we have to re`partition!` the mesh inside the function `p4est_partition_allow_for_coarsening!`?

Reply:
That depends on when this function is called. If it is called between mesh creation and creation of the semidiscretization then we do not have to repartition here (as that happens automatically during cache creation). But if it is called at some later stage then everything (including cache etc.) has to be repartitioned/reinitialized, not just the mesh.

Is the function still needed now that the AMR callback does not override anything anymore? If not, I'd suggest to simply delete it.

Reply:
Yes, let's delete it. It should just be set at the beginning to avoid all of these issues.

Reply:
Yes, sounds reasonable. Thanks.

Reply:
Done in f87f31abf7d3f495f9c501f2a38d23554e0fd960.

Reply:
Great! Thanks again to both of you for noticing this potential issue and resolving it so quickly!